accepted/2023/simplify-workload-versioning.md

```diff
@@ -0,0 +1,140 @@
+
+```
+> dotnet --version
+8.0.201
+```
+
+We will add a new `dotnet workload --version` command to print the workload patch version:
+
+```
+> dotnet workload --version
+8.0.201.2
+```
+
+We will also update `dotnet --info` and `dotnet workload --info` to display the workload patch version.  For example:
+
+```
+> dotnet --info
+.NET SDK:
+ Version:              8.0.201
+ Commit:               <commit>
+ Workloads version:    8.0.201.2
+
+<further information>
+
+> dotnet workload --info
+ Workloads version:     8.0.201.2
+ [wasm-tools]
+   Installation Source: SDK 8.0.100-preview.4
+   Manifest Version:    8.0.0-preview.4.23181.9/8.0.100-preview.4
+   Manifest Path:       C:\git\dotnet-sdk-main\.dotnet\sdk-manifests\8.0.100-preview.4\microsoft.net.workload.mono.toolchain.current\WorkloadManifest.json
+   Install Type:        FileBased
+```
+
+When updating workloads, console output will include the old and new workload patch versions:
+
+```
+> dotnet workload install wasm-tools
+Checking for updated workloads version...
+Updating workloads version from 8.0.201.0 to 8.0.201.2...
+Installing workload manifest microsoft.net.sdk.android version 34.0.0-preview.4.230...
+<Further workload manifest update messages>
+Installing pack Microsoft.NET.Runtime.WebAssembly.Sdk version 8.0.0-preview.4.23181.9...
+<Further workload pack installation manifests>
+Garbage collecting for SDK feature band(s) 8.0.200...
+
+Successfully updated workloads version from 8.0.201.0 to 8.0.201.2.
+Successfully installed workload(s) wasm-tools.
+```
+
+The proposed [workload history](https://github.com/dotnet/sdk/pull/30486) command will also display the workload patch versions.
+
+The .NET SDK will periodically (once a day, by default) check for updated workload versions in order to notify users if there is an update available.  Currently, commands such as `dotnet build` print out the following message:
+
+```
+Workload updates are available. Run `dotnet workload list` for more information.
+```
+
+The `dotnet workload list` command, in turn, prints out the following message:
+
+```
+Updates are available for the following workload(s): wasm-tools. Run `dotnet workload update` to get the latest.
+```
+
+We will modify both commands to print the same message:
+
+```
+A workload update (version 8.0.201.2) is available. Run `dotnet workload update` to install the latest workloads.
+```
+
+We will add a new `--version` option to `dotnet workload update` to allow installing (or downgrading to) a specific workloads version:
+
+```
+> dotnet workload update --version 8.0.201.0
+Updating workloads version from 8.0.201.2 to 8.0.201.0...
+<Manifest and pack installation / garbage collection messages>
+Successfully updated workloads version from 8.0.201.2 to 8.0.201.0.
+```
+
+## Specifying workload versions with global.json
+
+We will add support for specifying the workloads version in global.json.  For example:
```

Comment:
Wouldn't this break global.json? It would look for an SDK with that version, which would never exist, or is the plan to look for an SDK version of 8.0.201 and then a workload patch version of 8.0.201.0?

How would someone specify a patch version when the SDK version is prerelease? We cannot have a 4-part version number as that breaks semver 2.0

Should we consider adding a separate entry instead, e.g. `"workload-patch": "0"? That still won't solve the problem with preview SDKs I think

Reply:
The idea was that only the first three parts of the version would be used for SDK selection.  You're right though, it's probably better to have a separate value for the workloads version though.